lib/jwt/decode.rb

```diff
@@ -32,12 +34,15 @@ def decode_segments
 
     private
 
+    def verify_algo
+      raise(JWT::IncorrectAlgorithm, 'An algorithm must be specified') if allowed_algorithms.empty?
+      raise(JWT::IncorrectAlgorithm, 'Expected a different algorithm') unless options_includes_algo_in_header?
+    end
+
     def verify_signature
       @key = find_key(&@keyfinder) if @keyfinder
       @key = ::JWT::JWK::KeyFinder.new(jwks: @options[:jwks]).key_for(header['kid']) if @options[:jwks]
-
-      raise(JWT::IncorrectAlgorithm, 'An algorithm must be specified') if allowed_algorithms.empty?
-      raise(JWT::IncorrectAlgorithm, 'Expected a different algorithm') unless options_includes_algo_in_header?
+      @key = JWT::X5cKeyFinder.from(header['x5c'], @options.fetch(:root_certificates), @options[:crls]) if header['x5c']
```

Comment:
Is there a reason why `@options.fetch(:root_certificates)` and not `@options[:root_certificates]`?

Reply:
A `nil` value for root certificates doesn't make sense - it's not an optional value unlike CRLs. Raising an error here is more clear it's being used wrong than in `X5cKeyFinder.build_store`.

Reply:
Refactored this to happen inside the keyfinder class.